src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,368 @@
+import optax
+from .dataset import Dataset
+from .domain import Domain
+
+
+
+#######################################################
+#                                                     #
+#       Here is a the projectable data class          #
+#                                                     #
+#######################################################
+
+class ProjectableData():
+    def __init__(
+            self, 
+            D: jax.Array, 
+            domain: Domain
+        ):
+        """
+        A class storing one-hot data, supporting projection, vector 
+        query, and synthesize data operation.
+        
+        Args:
+            D: a jax.Array, representing a one-hot form of the data.
+            domain: a Domain class.
+        """
+        self.D_prime = D
+        self.domain = domain
+        self.feature_indices_map = {}
+        s = 0
+        for col in self.domain.attrs:
+            e = s + self.domain.size(col)
+            self.feature_indices_map[col] = list(range(s, e))
+            s = e
+
+    def datavector(self):
+        splits = [self.D_prime[:, self.feature_indices_map[attr]] for attr in self.domain.attrs]
+
+        letters = "bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        assert(
+            len(letters) >= len(self.domain)
+        ), 'High dim domain'
+
+        input_dims = ','.join(f'a{letters[i]}' for i in range(len(splits)))
+        output_dims = ''.join(f'{letters[i]}' for i in range(len(splits)))
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).mean(axis=0)
+        return np.array(joint).flatten()
+
+    def project(self, cl: tuple):
+        cols = list(itertools.chain.from_iterable(self.feature_indices_map[attr] for attr in cl))
+        D_proj = self.D_prime[:, cols]
+        domain_proj = self.domain.project(cl)
+        return ProjectableData(D_proj, domain_proj)
+    
+    def synthetic_data(self, num_samples, seed=0):
+        n, _ = self.D_prime.shape
+        key = jax.random.PRNGKey(seed)
+        key, sk = jax.random.split(key)
+        sampled_indices = jax.random.choice(key, n, shape=(num_samples,), replace=True)
+        sampled_data = self.D_prime[sampled_indices]
+
+        result = []
+        for attr in self.domain.attrs:
+            p = sampled_data[:, self.feature_indices_map[attr]]
+            key, sk = jax.random.split(key)
+            ids = jax.random.categorical(sk, jnp.log(jnp.clip(p, a_min=1e-30, a_max=1.0)), axis=-1)
+            result.append(ids)
+
+        sampled_df = pd.DataFrame(np.stack([np.array(r) for r in result], axis=1), columns = list(self.domain.attrs))
+        sampled_data = Dataset(sampled_df, self.domain)
+        return sampled_data
+
+
+def relaxed_projection_estimation(
+    domain: Domain,
+    measurements: List[tuple],
+    iters: int = 1000,
+    lr: float = 0.01,
+    D_start: jax.Array = None,
+    seed: int = 0,
+    D_size: int = 1000,
```

Comment:
Can we either make this a required kwarg or have it be optional and default to None (in which case it will be estimated from the measurements)